---
title: "Categorical Sequence Analysis with Optimal Matching: An Application with Data from the 'Women 40+ Healthy Aging Study'"
author: 
  - Adriana Clavijo Daza\newline
  - \newline
  - Supervised by:\newline
  - Prof. Dr. David Ginsbourger\newline
  - Dr. Serena Lozza-Fiacco
institute:
  - Statistics and Data Science Master's, Universität Bern
date: "2022-06-02"
header-includes:
  - \usepackage{algpseudocode}
output: 
  beamer_presentation:
    keep_tex: true
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  warning = FALSE,
  ## Default width x height for figures for LaTex is 10 x 7. Keep the ratio but
  ## reduce it by 25% making larger plots for presentation. Individual plots can
  ## be adjusted as needed.
  fig.width = 7.5,
  fig.height = 5.25
)
```

```{r libs, echo = FALSE}
library(readxl)
library(tidyr)
library(dplyr)
library(stringr)
library(ggplot2)
library(TraMineR)
library(TraMineRextras)
library(cluster)

## Note: this is order dependent.
source("../helpers/distance_matrix.R", local = knitr::knit_global())
source("../helpers/predict_scores.R", local = knitr::knit_global())
source("../helpers/get_MSE.R", local = knitr::knit_global())
```


## Motivation

```{r, out.width = "150px", fig.align = "center"}
knitr::include_graphics("../../Input/couple.png")
```

Understand the similarities and differences in the romantic relationships history of a group of women over 40 years old and explore the use of this information as a predictor for other psychosocial characteristics of interest.


## Women 40+ Healthy Aging Study (i)

```{r, out.width = "38%", out.height = "15%", fig.show = "hold", fig.align = "center"}
knitr::include_graphics(c("../../Input/logo2.png", "../Input/logo1.png"))
``` 

- Data from 250 women (ages 40-75) collected between June 2017 and February 2018.
- Information about relationship phases starting from the age of 15 years until the current age obtained with a psychometric instrument.
- Phases defined by the start and end age. 
- For each phase: civil status, relationship status, living situation, children and quality of the relationship.
- Additional information collected, in particular, scores for _personality traits_.


## Obtaining the sequences

Consider the relationship status: no relationship (NR), in a relationship (R), open relationship (OR), changing relationships (CR).

```{r, out.width = "280px"}
knitr::include_graphics("../../Input/sequence_example1.png")
```
*Current age

- Manual and automatic checks to identify inconsistent and incomplete records. Corrections based on a secondary data source.
- Identification and selection of the variables that provide a wider perspective of the relationship situation at a given time.
- Sequence data available for 239 individuals.


## Considered states

- 1 = Single + no children
- 2 = Single + children
- 3 = Changing relationships + no children
- 4 = Changing rel. + children
- 5 = Relationship + living apart + no children
- 6 = Relationship + living together + no children 
- 7 = Relationship + living apart + children
- 8 = Relationship + living together + children
- 9 = Married + no children
- 10 = Married + children

```{r, out.width = "280px"}
knitr::include_graphics("../../Input/sequence_example2.png")
```


## Overall distribution of states

```{r, child = "../data_prep.Rmd"}
```

```{r final-states, fig.align = "center", out.width = "300px"}
plot5 <- ggplot(data = rh_data, aes(x = Status_char))
plot5 +
  geom_bar() +
  geom_text(stat = "count", aes(label = ..count..), vjust = -0.25) +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme(axis.text = element_text(size = 10)) +
  labs(x = "State", y = "Count")
```


## Distribution of states by year

```{r, child = "../main_analysis.Rmd"}
```

```{r plot-seqdplot-0, out.width = "300px", fig.align = "center"}
seqdplot(results$sd,
  sortv = "from.start",
  ltext = status_labels,
  border = NA,
  cex.legend = 0.55
)
```


## What is personality?

```{r, out.width = "200px", fig.align = "center"}
knitr::include_graphics("../../Input/personality1.png")
```

Personality refers to the enduring characteristics and behavior that comprise a person’s unique adjustment to life, including major traits, interests, drives, values, self-concept, abilities, and emotional patterns.
<!-- TODO(): add citation/reference. -->

## The "Big Five" personality traits

```{r, out.width = "300px", fig.align = "center"}
knitr::include_graphics("../../Input/personality2.png")
```


## Research question

```{r, out.width = "280px", fig.align = "center", results = "hide"}
seqplot.tentrop(res, group = temp$Neuroticism_level, ylim = c(0, 1))
```

- Can we get a good prediction of personality scores based on the relationship history sequences?


## Optimal Matching (OM)

- Technique used in social sciences for the comparison of sequences of categorical states indexed by time.
- Applications on life course and career path analysis. 
- Uses the Needleman-Wunsch algorithm, that was developed to compare biological sequences.
- The Needleman-Wunsch algorithm is an application of dynamic programming, an iterative method that simplifies an optimization problem by breaking it into a recursion of smaller problems.


## The OM algortihm (i)

- Set of $n$ states: $S = \{s_1, \dots, s_n\}$.
- Sequence of size $t > 0$: $X = (x_1, \dots, x_t)$, with $x_i \in S$ for $i = 1, \dots, t$. 
- $\mathbf{S}$ is the set of all possible sequences with states belonging to $S$.

*Objective:* find the optimal way to align these two sequences.

- Let $X, Y \in \mathbf{S}$ be two sequences of size $t_1$ and $t_2$, respectively. 
- Define an empty array $F$ of size $(t_1+1) \times (t_2+1)$.


## The OM algortihm (ii)

\begin{algorithmic}[1]
\State $F(1, 1) \gets 0$
\For{$j \gets 2,t_2+1$}
  \State $F(1,j) \gets F(1, j-1) + d$
\EndFor
\For{$i \gets 2,t_1+1$}
  \State $F(i,1) \gets F(i-1, 1) + d$
\EndFor
\For{$i \gets 2,t_1+1$}
  \For{$j \gets 2,t_2+1$}
    \State $F(i,j) \gets \min\{F(i-1, j)+d, F(i, j-1)+d, F(i-1, j-1)+k(y_{i-1}, x_{j-1})\}$
    \EndFor
\EndFor
\end{algorithmic}


## The OM algortihm (iii)

- $d$ is the cost of inserting a gap (indel cost). 
- $k(y_{i-1}, x_{j-1})$ is the cost associated to change from the state $y_{i-1}$ to $x_{j-1}$.
- These costs are defined in a matrix $K$ of size $n \times n$ $\rightarrow$ cost matrix.
- Lines 1-7 of the algorithm correspond to initialization.
- The remaining lines of the algorithm correspond to the row-wise recursion to fill the array $F$.
- When $F$ is completely filled, the value $F(t_1+1, t_2+1)$ corresponds to the optimal cost of aligning the sequences $X$ and $Y$.


## Cost matrix (i)

The `R` package `TraMineR` provides several functions to work with sets of sequences. The package implements OM and offers several methods for computing the cost matrix $K$.


## Cost matrix (ii)

- Transition rates (`TRATE`):

The substitution cost between states $s_i$ and $s_j$, $1 \leq i, j \leq n$, is calculated as:

\begin{equation}
\label{eq:transition}
K(s_i, s_j) = c - P(s_i|s_j) - P(s_j|s_i),
\end{equation}

where $P(s_i|s_j)$ is the probability of transition from state $s_j$ in time $t$ to $s_i$ in time $t+1$ and $c$ is a constant, set to a value such that $0 \leq K(s_i, s_j) \leq 2$).


## Cost matrix (iii)

- Chi-squared distance (`FUTURE`):

\begin{equation}
\label{eq:chisq}
K(s_i, s_j) = d_{\chi^2}(\mathbf{P_i}, \mathbf{P_j}),
\end{equation}

where $\mathbf{P_.} = (P(s_1|s_.), \dots, P(s_n|s_.))'$


## Cost matrix (iv)

- Relative frequencies (`INDELS` and `INDELSLOG`):

\begin{equation}
\label{eq:indels}
K(s_i, s_j) = indel_i + indel_j,
\end{equation}

where $indel_i = 1/f_i$ for method `INDEL`, $indel_i = \log[2/(1+f_i)]$ and $f_i$ is the relative frequency of the state $s_i$ for $i = 1, \dots, n$. 


## Example (i)

Setup:

- $S$ = the alphabet
- $X = \{S,E,N,D\}, Y = \{A,N,D\} \in \mathbf{S}$
- $d = 2$  

$$
K(i,j) = 
\begin{cases}
0 & \text{if } i=j,\\
3 & \text{otherwise}
\end{cases}
$$


## Example (ii)

Initialization of $F$:

|   |   | S | E | N | D |
|---|---|---|---|---|---|
|   | 0 | 2 | 4 | 6 | 8 |
| A | 2 |   |   |   |   |
| N | 4 |   |   |   |   |
| D | 6 |   |   |   |   |


- $F(2,2) = \min\{F(1, 2)+d, F(2, 1)+d, F(1, 1)+k(y_{1}, x_{1})\} = \min\{2+2, 2+2, 0+3\} = 3$
- $F(2,3) = \min\{F(1, 3)+d, F(2, 2)+d, F(1, 2)+k(y_{1}, x_{2})\} = \min\{4+2, 3+2, 2+3\} = 5$
- $F(2,4) = \min\{F(1, 4)+d, F(2, 3)+d, F(1, 3)+k(y_{1}, x_{3})\} = \min\{6+2, 5+2, 4+3\} = 7$
- $F(2,5) = \min\{F(1, 5)+d, F(2, 4)+d, F(1, 4)+k(y_{1}, x_{4})\} = \min\{8+2, 7+2, 6+3\} = 9$


## Example (iii)

|   |   | S | E | N | D |
|---|---|---|---|---|---|
|   | 0 | 2 | 4 | 6 | 8 |
| A | 2 | 3 | 5 | 7 | 9 |
| N | 4 |   |   |   |   |
| D | 6 |   |   |   |   |


- $F(3,2) = \min\{F(2, 2)+d, F(3, 1)+d, F(2, 1)+k(y_{2}, x_{1})\} = \min\{3+2, 4+2, 2+3\} = 5$
- $F(3,3) = \min\{F(2, 3)+d, F(3, 2)+d, F(2, 2)+k(y_{2}, x_{2})\} = \min\{5+2, 5+2, 3+3\} = 6$
- $\dots$


## Example (iv)

|   |   | S | E | N | D |
|---|---|---|---|---|---|
|   | 0 | 2 | 4 | 6 | 8 |
| A | 2 | 3 | 5 | 7 | 9 |
| N | 4 | 5 | 6 | 5 | 7 |
| D | 6 | 7 | 8 | 7 | 5 |

S E N D with  
A - N D or $\hspace{0.5cm}$ ---> $\hspace{0.5cm}$ Optimal (equivalent) alignments   
-\ A N D  

- 2 matches: 0 
- 1 mismatch: 3 
- 1 gap: 2
- Total: 5


## Normalization

Given $X, Y \in \mathbf{S}$ of length $t_1$ and $t_2$, respectively. 
Let $d(X,Y)$ be the distance between the sequences $X$ and $Y$, $t_{max}$ the length of the longest sequence in $\mathbf{S}$ and $d_{max}$ the maximum distance between any pair of sequences in $\mathbf{S}$.

`TraMineR` provides the following options for normalization:

- `maxlength`: $d(X,Y)/t_{max}$
- `gmean`: $1- \frac{d_{max}-d(X,Y)}{\sqrt{t_1*t_2}}$
- `maxdist`: $d(X,Y)/d_{max}$


## Cost matrix

Base setup: `TRATE` and `maxlenght`

```{r cost_matrix}
knitr::kable(cost_matrix_base, digits = 2, row.names = FALSE) %>%
  ## Default font size is 11; reducing it to easily fit in slide. See:
  ## bookdown.org/yihui/rmarkdown-cookbook/kableextra.html#set-the-font-size
  kableExtra::kable_styling(font_size = 8)
```

```{r plot-dendogram, out.width = "300px", fig.align = "center", eval = FALSE}
## Application: clustering
plot(cluster_ward, labels = FALSE, hang = -1, xlab = "", ylab = "", sub = "")
abline(h = 5.5, col = "red")
```

## Clustering 

### State distribution by cluster

```{r plot-seqdplot-1, out.width = "300px", fig.align = "center"}
# Reduce the bottom and right margin of (sub)plots by ~50%.
par(mar = par()$mar * c(0.5, 1, 1, 0.5))

seqdplot(results$sd,
  group = clusters4_labels,
  border = NA,
  ltext = status_labels
)
```


## Application: clustering

### Transversal entropy by cluster

```{r, out.width = "280px", fig.align = "center", results = "hide"}
seqplot.tentrop(results$sd, 
                group = clusters4_labels, 
                ylim = c(0, 1),
                main = "")
```


## Clustering
### Characterization of the clusters

```{r}
knitr::kable(clusters4_counts) %>%
  kableExtra::kable_styling(position = "center")
```

&nbsp;

- Cluster 1: Married young and had children.
- Cluster 2: Often in relationships but not married.
- Cluster 3: Older, mostly married or in long relationship without children.
- Cluster 4: Younger, single or in a relationship without children.


## Clustering

### Personality scores distribution by cluster

```{r, out.width = "300px", fig.align = "center"}
p <- ggplot(data_all_long, aes(x = Trait_value, after_stat(density)))
p +
  geom_histogram(bins = 12) +
  facet_grid(
    rows = vars(Cluster_4),
    cols = vars(Trait),
    scales = "free_x"
  ) +
  labs(x = "Trait Value", y = "Density")
```


## Clustering

### State distribution for two clusters

```{r plot-seqdplot-2, out.width = "300px", fig.align = "center"}
seqdplot(results$sd,
  group = clusters2_labels,
  border = NA,
  ltext = status_labels,
  cex.legend = 0.75
)
```


## Application: clustering

### Transversal entropy for two clusters

```{r, out.width = "280px", fig.align = "center", results = "hide"}
seqplot.tentrop(results$sd, group = clusters2_labels, ylim = c(0, 1))
```


## Application: clustering

### Personality scores distribution for two clusters

```{r, fig.width = 7.5, fig.height = 5.25, out.width = "300px", fig.align = "center"}
p <- ggplot(data_all_long, aes(x = Trait_value, after_stat(density)))
p +
  geom_histogram(bins = 10) +
  facet_grid(
    rows = vars(Cluster_2),
    cols = vars(Trait),
    scales = "free_x"
  ) +
  labs(x = "Trait Value", y = "Density")
```


## Descriptive statistics for personality scores

```{r}
knitr::kable(pers_descrip, digits = 2) %>%
  kableExtra::kable_styling(font_size = 8)
```

&nbsp;

```{r, fig.height = 3, fig.width = 7.5, out.width = "300px", fig.align = "center"}
p <- ggplot(data_all_long, aes(x = Trait_value, after_stat(density)))
p +
  geom_histogram(bins = 10) +
  facet_grid(
    cols = vars(Trait),
    scales = "free_x"
  ) +
  labs(x = "Trait Value", y = "Density")
```


## $k$-nearest neighbors ($k$NN)

- Non-parametric method for prediction.
- Training set $\mathcal{D} = {(x_1, y_1), (x_2, y_2), \dots, (x_n, y_n)}$ of $n$ labeled data points, where $x_i \in \mathbb{R}^d$ and $y_i \in \mathcal{Y}$ (a finite set of for classification or a continuous range of values for regression). 
- Predict the label or value ($y_{n+1}$ unknown) for $x_{n+1}$ by finding the $k$ training data points closest to $x_{n+1}$ and taking a majority vote of their labels (for classification) or averaging the values of $Y$ (for regression).
- To compare the performance of different values of $k$, we calculate the mean squared error (MSE) in a test set of size $m$.

$$\text{MSE} = \frac{1}{m} \sum_{i=1}^n (y_i - \hat{y}_i)^2$$

where $y_i$ is the observed value and $\hat{y}_i$ is the predicted value via $k$NN.


## Prediction of personality scores (i)

- Only individuals who have available personality scores.
- New sample size: 200 individuals. 
- Data split into two subsets: train (70%) and test (30%). 
- Evaluate the MSE of the predictions in the test set only using the data from the nearest neighbors available in the train set.


## Prediction of personality scores (ii)

```{r, fig.width = (7.5 * 0.8), fig.height = (5.25 * 1.25),  out.width = "220px", fig.align = "center"}
base_MSE %>%
  filter(k > 3) %>%
  ggplot() +
  geom_line(aes(x = k, y = MSE)) +
  geom_line(
    aes(x = k, y = Trivial, col = "red"),
    alpha = 0.5,
    show.legend = FALSE
  ) +
  facet_wrap(
    facets = vars(Score),
    ncol = 1,
    scales = "free",
    strip.position = "top"
  ) +
  theme_minimal() +
  theme(strip.text = element_text(hjust = 0))
```


## Prediction of personality scores (iii)

- Method for calculation of cost matrix.
- Constant $c$ for cost matrix obtained from transition rates.
- Normalization of distances.
- Cost of transition from/to NA. 
- Limit range of the sequence.


## Prediction of personality scores

### Scenarios considered

```{r}
rownames(experiments) <- 1:nrow(experiments)
knitr::kable(experiments, digits = 2, row.names = TRUE) %>%
  ## Default font size is 11; reducing it to easily fit in slide. See:
  ## bookdown.org/yihui/rmarkdown-cookbook/kableextra.html#set-the-font-size
  kableExtra::kable_styling(font_size = 8)
```


## Prediction of personality scores

### Best prediction by scenarios and Trait

```{r, out.width = "325px", fig.align = "center"}
my_exp %>%
  ggplot() +
  aes(Scenario, Score) +
  geom_tile(aes(fill = Improvement)) +
  geom_text(
    aes(label = paste(round(Improvement, 2), paste0("k = ", k), sep = "\n")),
    size = 2.25, # default is GeomLabel$default_aes$size (3.88)
  ) +
  scale_fill_gradient(
    low = rgb(1, 0, 0, alpha = 0.7),
    high = rgb(0, 1, 0, alpha = 0.7)
  ) +
  # A recommended alternative color scale is:
  # scale_fill_viridis_c(alpha = 0.7)
  scale_x_continuous(breaks = 1:11) +
  coord_fixed(ratio = 1) +
  labs(y = "Trait") +
  theme_minimal() +
  theme(panel.grid = element_blank())
```


## Neuroticism - Scenario 3

```{r, out.width = "250px", fig.align = "center"}
eval_experiments %>%
  tibble::add_column(Scenario = 1:12) %>%
  select(Scenario, result) %>%
  tidyr::unnest(result) %>%
  filter(
    Scenario == 3,
    Score == "Neuroticism",
    k > 1
  ) %>%
  ggplot() +
  geom_line(aes(x = k, y = MSE)) +
  geom_line(
    aes(x = k, y = Trivial, col = "red"),
    alpha = 0.5,
    show.legend = FALSE
  ) +
  theme_minimal()
```


## Neuroticism - Scenario 7

```{r, out.width = "250px", fig.align = "center"}
eval_experiments %>%
  tibble::add_column(Scenario = 1:12) %>%
  select(Scenario, result) %>%
  tidyr::unnest(result) %>%
  filter(
    Scenario == 7,
    Score == "Neuroticism",
    k > 1
  ) %>%
  ggplot() +
  geom_line(aes(x = k, y = MSE)) +
  geom_line(
    aes(x = k, y = Trivial, col = "red"),
    alpha = 0.5,
    show.legend = FALSE
  ) +
  theme_minimal()
```


## Openness

```{r, out.width = "250px", fig.align = "center"}
eval_experiments %>%
  tibble::add_column(Scenario = 1:12) %>%
  select(Scenario, result) %>%
  tidyr::unnest(result) %>%
  filter(
    Scenario == 12,
    Score == "Openness",
    k > 1
  ) %>%
  ggplot() +
  geom_line(aes(x = k, y = MSE)) +
  geom_line(
    aes(x = k, y = Trivial, col = "red"),
    alpha = 0.5,
    show.legend = FALSE
  ) +
  theme_minimal()
```


## Extraversion

```{r, out.width = "250px", fig.align = "center"}
eval_experiments %>%
  tibble::add_column(Scenario = 1:12) %>%
  select(Scenario, result) %>%
  tidyr::unnest(result) %>%
  filter(
    Scenario == 8,
    Score == "Extraversion",
    k > 4
  ) %>%
  ggplot() +
  geom_line(aes(x = k, y = MSE)) +
  geom_line(
    aes(x = k, y = Trivial, col = "red"),
    alpha = 0.5,
    show.legend = FALSE
  ) +
  theme_minimal()
```


## Conscientiousness

```{r, out.width = "250px", fig.align = "center"}
eval_experiments %>%
  tibble::add_column(Scenario = 1:12) %>%
  select(Scenario, result) %>%
  tidyr::unnest(result) %>%
  filter(
    Scenario == 1,
    Score == "Conscientiousness",
    k > 4
  ) %>%
  ggplot() +
  geom_line(aes(x = k, y = MSE)) +
  geom_line(
    aes(x = k, y = Trivial, col = "red"),
    alpha = 0.5,
    show.legend = FALSE
  ) +
  theme_minimal()
```


## Agreeableness

```{r, fig.width = 7.5, fig.height = 5.25, out.width = "250px", fig.align = "center"}
eval_experiments %>%
  tibble::add_column(Scenario = 1:12) %>%
  select(Scenario, result) %>%
  tidyr::unnest(result) %>%
  filter(
    Scenario == 12,
    Score == "Agreeableness",
    k > 4
  ) %>%
  ggplot() +
  geom_line(aes(x = k, y = MSE)) +
  geom_line(
    aes(x = k, y = Trivial, col = "red"),
    alpha = 0.5,
    show.legend = FALSE
  ) +
  theme_minimal()
```



## Final comments

- OM provides a way to use categorical sequences for prediction.
- In most scenarios, the configuration of the clusters remained unchanged.
- The distance matrix highly sensitive to the definition of the cost matrix.
- Symmetry of the cost matrix may not be appropriate.
- The quality of the prediction is affected by the method chosen to handle missing values.
- The methodology makes it difficult to include additional variables for prediction.


## References

- Optimal Matching Methods for Historical Sequences - A. Abbott & J. Forrest (1986).
- A General Method Applicable to the Search for Similarities in the Amino Acid Sequence of Two Proteins - S. Needleman & C. Wunsch (1970).
- Sequence Analysis: New Methods for Old Ideas - A. Abbott (1995).
- Optimal Matching Analysis: A Methodological Note on Studying Career Mobility - T. W. Chan (1995).
- Analyzing Sequence Data: Optimal Matching in
Management Research - T. Biemann & D. K. Datta (2013).
- Analyzing and Visualizing State Sequences in R
with TraMineR - A. Gabadinho, G. Ritschard, N. S. Müller, M. Studer (2011).
